cpp/lib.py

```diff
@@ -1204,7 +1199,11 @@ def run_task_on_codebase(self, task: str, languages=None, sources=None, **kwargs
             cmd += src_dirs
             log_command(cmd)
             if not sources or src_dirs:
-                git_ls_tree = subprocess.check_output(cmd).decode("utf-8").split("\0")
```

Comment:
This change is quite delicate as there are many uses cases (I need to document / test this)

let me highlight a use-case that this change breaks with a concrete example within the STEPS repository

### STEPS repository has C++ files in the directory `src/steps/math/`

```
$ cd /path/to/STEPS-WORKING_COPY
$ ls src/steps/math/
CMakeLists.txt	distributions.hpp  ghk.hpp	 point.hpp	sphere.cpp	 tetrahedron.hpp  triangle.cpp
bbox.hpp	doxygen.hpp	   linsolve.cpp  sample.hpp	sphere.hpp	 tools.cpp	  triangle.hpp
constants.hpp	ghk.cpp		   linsolve.hpp  smallsort.hpp	tetrahedron.cpp  tools.hpp
```

### The `format` utility can be given input files or directories relative to CWD

```
$ cd src/steps
$ ../../CMake/hpc-coding-conventions/bin/format --lang c++ math
/usr/local/opt/llvm/bin/clang-format -i src/steps/math/linsolve.hpp src/steps/math/sphere.hpp src/steps/math/tools.hpp src/steps/math/tetrahedron.cpp src/steps/math/ghk.cpp src/steps/math/sample.hpp src/steps/math/triangle.cpp src/steps/math/point.hpp src/steps/math/ghk.hpp src/steps/math/bbox.hpp src/steps/math/distributions.hpp src/steps/math/tetrahedron.hpp src/steps/math/sphere.cpp src/steps/math/smallsort.hpp src/steps/math/tools.cpp src/steps/math/triangle.hpp src/steps/math/doxygen.hpp src/steps/math/constants.hpp src/steps/math/linsolve.cpp
```


## With this change

```
$ ../../CMake/hpc-coding-conventions/bin/format --lang c++ math
$
```

=> git_ls_tree doesn't find `math` at repository root directory.



As I said, this part of the code handles too many situations with too many conditional statements...
Can you please describe the use-case that requires you to change this @olupton ?

Reply:
Without this change then `ninja clang-tidy` does not find any files. That target changes to a directory buried inside the CMake build directory before running the `clang-tidy` wrapper:
```
COMMAND = cd /Users/olupton/build/nmodl/build/cmake/hpc-coding-conventions/cpp && /opt/local/bin/python3 /Users/olupton/build/nmodl/cmake/hpc-coding-conventions/cpp/../bin/clang-tidy -v -p /Users/olupton/build/nmodl/build/compile_commands.json
```
and running `git ls-tree` anywhere underneath the build directory does not yield anything. To pick up all files committed to git then [the current git command] needs to be run from the root of the git repository.

Reply:
This patch should fix the `ninja clang-tidy` issue. Would that suit you?

```diff
diff --git a/cpp/CMakeLists.txt b/cpp/CMakeLists.txt
index 8af39d9..55da552 100644
--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -64,7 +64,10 @@ function(cpp_cc_setup_tool_config name path)
 endfunction()

 function(cpp_cc_add_tool_target name)
-  add_custom_target(${name}-${PROJECT_NAME} ${ARGN})
+  add_custom_target(
+    ${name}-${PROJECT_NAME}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    ${ARGN})
   if(TARGET ${name})
     add_dependencies(${name} ${name}-${PROJECT_NAME})
   else()
```

Reply:
I think that's an improvement; I pushed the change!